### PR TITLE
Enable users to provide per-volume Hadoop Filesystem overrides

### PR DESCRIPTION
**Pull request by dlmarion**
When creating Hadoop FileSystem objects the FileSystem class caches FileSystem objects using the filesystem URI scheme and authority fields. Prior to this change, the Accumulo VolumeManager would create FileSystem objects (one per volume) using the standard Hadoop Configuration files and resources. However, user-supplied Hadoop configuration files (core-site.xml and hdfs-site.xml) don't allow the user to specify different properties for different filesystems. For example, a user can create a core-site.xml and hdfs-site.xml file locally and create FileSystem objects that connect to different HDFS clusters. A property specified in the Hadoop configuration will apply to all FileSystem objects.

With this change users can specify Hadoop configuration overrides in the Accumulo configuration using the property format: `instance.volumes.config.<volume-uri>.<property-name>=<property-value>`

This enables a user with multiple filesystems specified in the instance.volumes property to provide different Hadoop configuration overrides for each volume. For example, the following would allow different hedged read thread sizes for two different volumes:

instance.volumes=hdfs://namespace-a:8020/accumulo,hdfs://namespace-b:8020/accumulo instance.volumes.config.hdfs://namespace-a:8020/accumulo.dfs.client.hedged.read.threadpool.size=10 instance.volumes.config.hdfs://namespace-b:8020/accumulo.dfs.client.hedged.read.threadpool.size=20

Note: When specifying property names that contain colons in the accumulo.properties files, they need to be escaped with a backslash

### CODE REVIEW

**Comment by dlmarion** [comment]
Full IT build successful

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java [hunk 1]
```diff
@@ -353,6 +362,38 @@ public short getDefaultReplication(Path path) {
...
+   * @param conf AccumuloConfiguration object
+   * @param hadoopConf Hadoop Configuration object
+   * @param filesystemURI Volume Filesystem URI
+   * @return Hadoop Configuration with custom overrides for this FileSystem
+   */
+  private static Configuration getVolumeManagerConfiguration(AccumuloConfiguration conf,
+      final Configuration hadoopConf, final String filesystemURI) {
+    final Configuration volumeConfig = new Configuration(hadoopConf);
+    final Map<String,String> customProps =
+        conf.getAllPropertiesWithPrefixStripped(Property.GENERAL_ARBITRARY_PROP_PREFIX);
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:386]: I would use a different prefix for these properties.  I think of the general arbitrary prefix as something plugins use.  This config is not related to a plugin, its for core functionality.  Could do something like the following.

```
instance.volumes.dfs.config.<volume-uri>.<property-name>=<property-value>
```

This config prefix also makes the config sort together with the instance.volumes config.

File: test/src/main/java/org/apache/accumulo/test/VolumeManagerIT.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  }
+
+  @Test
+  public void testHdfsConfigOverrides() throws Exception {
+    try (AccumuloClient c = Accumulo.newClient().from(getClientProperties()).build()) {
+
+      Map<String,String> siteConfig = c.instanceOperations().getSiteConfiguration();
+      assertEquals("10485760", siteConfig
+          .get(Property.GENERAL_ARBITRARY_PROP_PREFIX.getKey() + vol1 + ".dfs.blocksize"));
+      assertEquals("104857600", siteConfig
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/VolumeManagerIT.java:90]: `104857600` is so similar to `10485760` that at first glance I thought both were the same.  Could replace `104857600` with something more visually distinct like `51200000`

File: server/base/src/test/java/org/apache/accumulo/server/fs/VolumeManagerImplTest.java [hunk 3]
```diff
@@ -104,4 +111,88 @@ public Scope getChooserScope() {
...
+    Configuration conf1 = fs1.getConf();
+
+    FileSystem fs2 = vm.getFileSystemByPath(new Path(vol2));
+    Configuration conf2 = fs2.getConf();
+
+    FileSystem fs3 = vm.getFileSystemByPath(new Path(vol3));
+    Configuration conf3 = fs3.getConf();
+
+    FileSystem fs4 = vm.getFileSystemByPath(new Path(vol4));
+    Configuration conf4 = fs4.getConf();
```
Comment (keith-turner) [inline server/base/src/test/java/org/apache/accumulo/server/fs/VolumeManagerImplTest.java:155]: Could add a 5th volume that does not set any overrides to ensure there is no interference.

File: server/base/src/test/java/org/apache/accumulo/server/fs/VolumeManagerImplTest.java [hunk 4]
```diff
@@ -104,4 +111,88 @@ public Scope getChooserScope() {
...
+    final String vol2 = "file://localhost/vol2/";
+    final String vol3 = "hdfs://127.0.0.1/accumulo";
+    final String vol4 = "hdfs://localhost/accumulo";
+
+    ConfigurationCopy conf = new ConfigurationCopy();
+    conf.set(Property.INSTANCE_VOLUMES, String.join(",", vol1, vol2, vol3, vol4));
+    conf.set(Property.GENERAL_VOLUME_CHOOSER, Property.GENERAL_VOLUME_CHOOSER.getDefaultValue());
+    conf.set(Property.GENERAL_ARBITRARY_PROP_PREFIX.getKey() + vol1 + "."
+        + HdfsClientConfigKeys.HedgedRead.THREADPOOL_SIZE_KEY, "10");
+    conf.set(Property.GENERAL_ARBITRARY_PROP_PREFIX.getKey() + vol1 + "."
```
Comment (keith-turner) [inline server/base/src/test/java/org/apache/accumulo/server/fs/VolumeManagerImplTest.java:128]: Could set a third override on one volume that no others set to ensure there is no interference for this case.  For example could set a third blocksize prop for only vol1.

**Review by keith-turner — COMMENTED** [review]
Not exactly sure where this should go in the code, but it would be nice to validate that the defined overrides are for a defined volumes.  For example in the config below volumes are defined for namespace-a and namespace-b.  Then an override is defined for namespace-z which is not a defined volume.  Checking for this could help catch user errors.

```
instance.volumes=hdfs://namespace-a:8020/accumulo,hdfs://namespace-b:8020/accumulo
general.custom.hdfs://namespace-a:8020/accumulo.dfs.client.hedged.read.threadpool.size=10
general.custom.hdfs://namespace-z:8020/accumulo.dfs.client.hedged.read.threadpool.size=20
```

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java [hunk 5]
```diff
@@ -353,6 +362,38 @@ public short getDefaultReplication(Path path) {
...
+    customProps.forEach((key, value) -> {
+      if (key.startsWith(filesystemURI)) {
+        String property = key.substring(filesystemURI.length() + 1);
+        log.debug("Overriding property {} to {} for volume {}", property, value, filesystemURI);
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java]: It should only happen once, at server startup. Increased logging level to INFO in e0c12c4.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java (hunk 1 above)
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java]: Created new property in e0c12c4.

File: test/src/main/java/org/apache/accumulo/test/VolumeManagerIT.java (hunk 2 above)
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/VolumeManagerIT.java]: Updated test to use proposed value in e0c12c4.

File: server/base/src/test/java/org/apache/accumulo/server/fs/VolumeManagerImplTest.java (hunk 3 above)
Comment (dlmarion) [reply server/base/src/test/java/org/apache/accumulo/server/fs/VolumeManagerImplTest.java:161]: Added a fifth volume to the test in e0c12c4.

File: server/base/src/test/java/org/apache/accumulo/server/fs/VolumeManagerImplTest.java (hunk 4 above)
Comment (dlmarion) [reply server/base/src/test/java/org/apache/accumulo/server/fs/VolumeManagerImplTest.java]: Added blocksize property to one volume in test.

**Comment by dlmarion** [comment]
@keith-turner - changes made based on your comments.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java [hunk 6]
```diff
@@ -353,11 +363,63 @@ public short getDefaultReplication(Path path) {
...
+
+    definedVolumes.forEach(vol -> {
+      log.debug("Looking for defined volume: {}", vol);
+      overrideProperties.keySet().forEach(override -> {
+        if (override.startsWith(vol)) {
+          log.debug("Found volume {}, removing property {}", vol, override);
+          overrideProperties.remove(override);
+        }
+      });
+    });
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:411]: This can be written much more succinctly. Plus, you can get false positives if you don't include the dot separator (for example, an override for `hdfs://host:port/path/to/accumulo2` will incorrectly be removed as a match against a volume named `hdfs://host:port/path/to/accumulo` if you don't include the dot to ensure the volume name is terminated).

```suggestion
    definedVolumes.forEach(
        vol -> overrideProperties.entrySet().removeIf(entry -> entry.getKey().startsWith(vol + ".")));
```

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java [hunk 7]
```diff
@@ -353,11 +363,63 @@ public short getDefaultReplication(Path path) {
...
+
+    overrideProperties.forEach((k, v) -> log
+        .warn("Found no matching volume for volume config override property {} = {}", k, v));
+  }
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:415]: On second thought, the entire method can be written without an intermediate map or dealing with any concurrency... just grab the properties you want to log a warning about.

```suggestion
  private static void warnVolumeOverridesMissingVolume(AccumuloConfiguration conf,
      Set<String> definedVolumes) {
    conf.getAllPropertiesWithPrefixStripped(Property.INSTANCE_VOLUMES_CONFIG).entrySet().stream()
        // log only configs where none of the volumes (with a dot) prefix its key
        .filter(e -> definedVolumes.stream().noneMatch(vol -> e.getKey().startsWith(vol + ".")))
        .forEach(
            e -> log.warn("Found no matching volume for volume config override property {}", e));
  }
```

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java [hunk 8]
```diff
@@ -353,11 +363,63 @@ public short getDefaultReplication(Path path) {
...
+    final Configuration volumeConfig = new Configuration(hadoopConf);
+    final Map<String,String> customProps =
+        conf.getAllPropertiesWithPrefixStripped(Property.INSTANCE_VOLUMES_CONFIG);
+    customProps.forEach((key, value) -> {
+      if (key.startsWith(filesystemURI)) {
+        String property = key.substring(filesystemURI.length() + 1);
+        log.info("Overriding property {} to {} for volume {}", property, value, filesystemURI);
+        volumeConfig.set(property, value);
+      }
+    });
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:394]: This can be written slightly more succinctly without an intermediate map variable. This doesn't save much, but I do think there's some expressive value in separating the Predicate `if` clause from the `forEach` action:

```suggestion
    conf.getAllPropertiesWithPrefixStripped(Property.INSTANCE_VOLUMES_CONFIG).entrySet().stream()
        .filter(e -> e.getKey().startsWith(filesystemURI + ".")).forEach(e -> {
          String key = e.getKey().substring(filesystemURI.length() + 1);
          String value = e.getValue();
          log.info("Overriding property {} for volume {}", key, value, filesystemURI);
          volumeConfig.set(key, value);
        });
```

And again, you need to include the dot when checking for the startsWith... or you'll get false positives from similarly named paths.

File: minicluster/src/main/java/org/apache/accumulo/miniclusterImpl/MiniAccumuloConfigImpl.java [hunk 9]
```diff
@@ -47,6 +47,12 @@
  * @since 1.6.0
  */
 public class MiniAccumuloConfigImpl {
+
+  @FunctionalInterface
+  public static interface FinalSiteConfigUpdater {
+    void update(MiniAccumuloConfigImpl cfg);
+  }
```
Comment (ctubbsii) [inline minicluster/src/main/java/org/apache/accumulo/miniclusterImpl/MiniAccumuloConfigImpl.java:54]: You don't need this interface. Just use `Consumer<MiniAccumuloConfigImpl>`. While it probably doesn't hurt to have an extra interface to select a more expressive name for ease of use, I think the name here, with the word "Final" here is a bit presumptive (what if we need a subsequent modification to support other tests? would we call it "FinalFinal"?). We can just avoid coming up with a name by just using the existing Consumer... also, Consumer has the helpful `andThen` method, which we don't need to reinvent and could be useful.

File: minicluster/src/main/java/org/apache/accumulo/miniclusterImpl/MiniAccumuloConfigImpl.java [hunk 10]
```diff
@@ -812,4 +820,29 @@ public int getNumCompactors() {
...
+  /**
+   * Set the FinalSiteConfigUpdater instance that will be used to modify the site configuration
+   * right before it's written out a file. This would be useful in the case where the configuration
+   * needs to be updated based on a property that is set in MiniAccumuloClusterImpl like
+   * instance.volumes
+   *
+   * @param updater FinalSiteConfigUpdater instance
+   * @since 2.1.1
+   */
+  public void setFinalSiteConfigUpdater(FinalSiteConfigUpdater updater) {
```
Comment (ctubbsii) [inline minicluster/src/main/java/org/apache/accumulo/miniclusterImpl/MiniAccumuloConfigImpl.java:833]: Even if the parameter is changed to a Consumer, this method still needs a decent name. I still have reservations about this one. One thing it has going for it is that it's not public API... it's only on the Impl class.

Still, maybe something like:
```suggestion
  public void preStartProcessor(Consumer<MiniAccumuloConfigImpl> prestartConfigProcessor) {
```

**Review by ctubbsii — CHANGES_REQUESTED** [review]
There's a few bugs in here that could arise with similarly named volumes where one volume is a prefix of another, so you need to check for the terminal dot at the end of the volume before the Hadoop config key name for that volume. Overall, though, I think this is a nice change.

There's some inconsistency in the naming that needs to be polished up also.

Aside from those, I made some suggestions regarding naming, since tags, and streamlined use of streams / forEach.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java [hunk 11]
```diff
@@ -353,11 +363,63 @@ public short getDefaultReplication(Path path) {
...
+      overrideProperties.keySet().forEach(override -> {
+        if (override.startsWith(vol)) {
+          log.debug("Found volume {}, removing property {}", vol, override);
+          overrideProperties.remove(override);
+        }
+      });
+    });
+
+    overrideProperties.forEach((k, v) -> log
+        .warn("Found no matching volume for volume config override property {} = {}", k, v));
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:414]: Would be really nice to unit test this case, however since it just logs a message not sure of the best way to do that.

Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java]: Implemented test in f34eb45.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java [hunk 12]
```diff
@@ -120,6 +120,16 @@ public enum Property {
...
+  INSTANCE_VOLUMES_CONFIG("instance.volumes.config.", null, PropertyType.PREFIX,
+      "Properties in this category are used to provide volume specific overrides to "
+          + "the general filesystem client configuration. Properties using this prefix "
+          + "should be in the form "
+          + "'instance.volumes.config.<volume-uri>.<property-name>=<property-value>. An "
+          + "example: "
+          + "'instance.volume.config.hdfs://namespace-a:8020/accumulo.dfs.client.hedged.read.threadpool.size=10'. "
+          + "Note that when specifying property names that contain colons in the properties "
+          + "files that the colons need to be escaped with a backslash.",
+      "2.1.1"),
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java]: Addressed in f34eb45.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java (hunk 6 above)
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java]: Addressed in f34eb45.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java (hunk 7 above)
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java]: Addressed in f34eb45.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java (hunk 8 above)
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java]: Addressed in f34eb45.

File: minicluster/src/main/java/org/apache/accumulo/miniclusterImpl/MiniAccumuloConfigImpl.java (hunk 9 above)
Comment (dlmarion) [reply minicluster/src/main/java/org/apache/accumulo/miniclusterImpl/MiniAccumuloConfigImpl.java]: Addressed in f34eb45.

**Comment by dlmarion** [comment]
@ctubbsii @keith-turner  - Thanks for the review, I believe that I implemented all of your suggested changes.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java [hunk 13]
```diff
@@ -120,11 +120,11 @@ public enum Property {
           + " a comma or other reserved characters in a URI use standard URI hex"
           + " encoding. For example replace commas with %2C.",
       "1.6.0"),
-  INSTANCE_VOLUMES_CONFIG("instance.volumes.config.", null, PropertyType.PREFIX,
+  INSTANCE_VOLUMES_CONFIG("instance.volume.config.", null, PropertyType.PREFIX,
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java:123]:
```suggestion
  INSTANCE_VOLUME_CONFIG("instance.volume.config.", null, PropertyType.PREFIX,
```

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java [hunk 14]
```diff
@@ -382,44 +382,37 @@ public short getDefaultReplication(Path path) {
...
 
-    overrideProperties.forEach((k, v) -> log
-        .warn("Found no matching volume for volume config override property {} = {}", k, v));
+  protected static List<Entry<String,String>>
+      findVolumeOverridesMissingVolume(AccumuloConfiguration conf, Set<String> definedVolumes) {
+    return conf.getAllPropertiesWithPrefixStripped(Property.INSTANCE_VOLUMES_CONFIG).entrySet()
+        .stream()
+        // log only configs where none of the volumes (with a dot) prefix its key
+        .filter(e -> definedVolumes.stream().noneMatch(vol -> e.getKey().startsWith(vol + ".")))
+        .collect(Collectors.toList());
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:405]: I see you needed this as a separate method for verification in the test. However, it's a bit weird to return a list of entries, rather than a map. But, there's also a performance hit by collecting it into a data structure in the first place. I don't think you need that at all for the test or for logging the warning. Just return the Stream, which you can call forEach on directly for the warning... and for the test you can check it as you stream, or call `.collect` on it there to do whatever verification you need.

```suggestion
  protected static Stream<Entry<String,String>>
      findVolumeOverridesMissingVolume(AccumuloConfiguration conf, Set<String> definedVolumes) {
    return conf.getAllPropertiesWithPrefixStripped(Property.INSTANCE_VOLUMES_CONFIG).entrySet()
        .stream()
        // log only configs where none of the volumes (with a dot) prefix its key
        .filter(e -> definedVolumes.stream().noneMatch(vol -> e.getKey().startsWith(vol + ".")));
```

**Comment by ctubbsii** [comment]
@dlmarion wrote:
> @ctubbsii @keith-turner - Thanks for the review, I believe that I implemented all of your suggested changes.

I just had two very minor follow-on suggested changes. After those are addressed, I'm good with merging.

**Review by keith-turner — APPROVED** [review]

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java (hunk 14 above)
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java]: Changes applied in 55fe4f3
